src/Grpc.NetCore.HttpClient/Internal/GrpcCall.cs

```diff
@@ -0,0 +1,246 @@
+                },
+                GrpcProtocolConstants.GrpcContentTypeHeaderValue);
+
+            var writer = new HttpContentClientStreamWriter<TRequest, TResponse>(this, writeStreamTcs.Task, completeTcs);
+            return writer;
+        }
+
+        private HttpRequestMessage CreateHttpRequestMessage()
+        {
+            var message = new HttpRequestMessage(HttpMethod.Post, Method.FullName);
+            message.Version = new Version(2, 0);
+
+            if (Options.Headers != null && Options.Headers.Count > 0)
+            {
+                foreach (var entry in Options.Headers)
+                {
+                    // Deadline is set via CallOptions.Deadline
+                    if (entry.Key == GrpcProtocolConstants.TimeoutHeader)
+                    {
+                        continue;
+                    }
+
+                    var value = entry.IsBinary ? Convert.ToBase64String(entry.ValueBytes) : entry.Value;
+                    message.Headers.Add(entry.Key, value);
+                }
+            }
+
+            if (Options.Deadline != null && Options.Deadline != DateTime.MaxValue)
+            {
+                var deadline = Options.Deadline.Value - _clock.UtcNow;
+
+                // JamesNK(todo) - Replicate C core's logic for formatting grpc-timeout
+                message.Headers.Add(GrpcProtocolConstants.TimeoutHeader, Convert.ToInt64(deadline.TotalMilliseconds) + "m");
+            }
+
+            return message;
+        }
+
+        public void EnsureNotDisposed()
+        {
+            if (Disposed)
+            {
+                throw new ObjectDisposedException(nameof(GrpcCall<TRequest, TResponse>));
+            }
+        }
+
+        public async Task<TResponse> GetResponseAsync()
+        {
+            _httpResponse = await SendTask.ConfigureAwait(false);
+            var responseStream = await _httpResponse.Content.ReadAsStreamAsync().ConfigureAwait(false);
+
+            var message = await responseStream.ReadSingleMessageAsync(Method.ResponseMarshaller.Deserializer, _callCts.Token).ConfigureAwait(false);
+            ResponseFinished = true;
+
+            // The task of this method is cached so there is no need to cache the message here
+            return message;
+        }
+
+        public async Task<Metadata> GetResponseHeadersAsync()
+        {
+            _httpResponse = await SendTask.ConfigureAwait(false);
+
+            // The task of this method is cached so there is no need to cache the headers here
+            return GrpcProtocolHelpers.BuildMetadata(_httpResponse.Headers);
+        }
+
+        public Status GetStatus()
+        {
+            ValidateTrailersAvailable();
+
+            var grpcStatus = SendTask.Result.TrailingHeaders.GetValues(GrpcProtocolConstants.StatusTrailer).FirstOrDefault();
+            var grpcMessage = SendTask.Result.TrailingHeaders.GetValues(GrpcProtocolConstants.MessageTrailer).FirstOrDefault();
```

Comment:
https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses:
>The value portion of Status-Message is conceptually a Unicode string description of the error, physically encoded as UTF-8 followed by percent-encoding.

We should unescape the perccent-encoding here.

Reply:
Added todo.